src/services/jsTyping.ts

```diff
@@ -0,0 +1,282 @@
+
+                    let nodeModulesPath = ts.combinePaths(searchDir, "node_modules");
+                    getTypingNamesFromNodeModuleFolder(nodeModulesPath, filesToWatch);
+                }
+
+                getTypingNamesFromSourceFileNames(fileNames);
+                getTypingNamesFromCompilerOptions(compilerOptions);
+            }
+
+            let newTsdJsonPath: string;
+            const typingsPath = ts.combinePaths(cachePath, "typings");
+            const tsdJsonPath = ts.combinePaths(cachePath, "tsd.json");
+            const tsdJsonDict = tryParseJson(tsdJsonPath);
+            if (tsdJsonDict) {
+                for (const notFoundTypingName of notFoundTypingNames) {
+                    if (inferredTypings.hasOwnProperty(notFoundTypingName) && !inferredTypings[notFoundTypingName]) {
+                        delete inferredTypings[notFoundTypingName];
+                    }
+                }
+
+                // The "installed" property in the tsd.json serves as a registry of installed typings. Each item 
+                // of this object has a key of the relative file path, and a value that contains the corresponding
+                // commit hash.
+                if (hasProperty(tsdJsonDict, "installed")) {
+                    for (const cachedTypingPath in tsdJsonDict.installed) {
+                        // Assuming the cachedTypingPath has the format of "[package name]/[file name]"
+                        const cachedTypingName = cachedTypingPath.substr(0, cachedTypingPath.indexOf('/'));
+                        // If the inferred[cachedTypingName] is already not null, which means we found a corresponding
+                        // d.ts file that coming with the package. That one should take higher priority.
+                        if (hasProperty(inferredTypings, cachedTypingName) && !inferredTypings[cachedTypingName]) {
+                            inferredTypings[cachedTypingName] = ts.combinePaths(typingsPath, cachedTypingPath);
+                        }
+                    }
+                }
+            } else if (!host.fileExists(tsdJsonPath) && typingOptions.enableAutoDiscovery) {
+                const tsdJsonOptions = {
+                    version: "v4",
+                    repo: "DefinitelyTyped/DefinitelyTyped",
+                    ref: "master",
+                    path: "typings",
+                };
+                host.writeFile(tsdJsonPath, JSON.stringify(tsdJsonOptions, null, "  "));
+                newTsdJsonPath = tsdJsonPath;
+            }
+
+            // Remove typings that the user has added to the exclude list
+            for (const excludeTypingName of exclude) {
+                delete inferredTypings[excludeTypingName];
+            }
+
+            const newTypingNames: string[] = [];
+            const cachedTypingPaths: string[] = [];
+            for (const typing in inferredTypings) {
+                if (inferredTypings[typing]) {

```

Comment:
It's not clear what you're testing here and why you're doing it. Are you testing for emptiness? Definedness?